pandas/core/indexing.py

```diff
@@ -1727,6 +1728,44 @@ def _getitem_axis(self, key, axis: int):
         self._validate_key(key, axis)
         return self._get_label(key, axis=axis)
 
+    def _ensure_listlike_indexer(self, key, is_indexer_key: bool):
+        """
+        Ensure that a list-like of column labels are all present by adding them if
+        they do not already exist.
+
+        Parameters
+        ----------
+        key : _LocIndexer key or list-like of column labels
+            Target labels.
+        is_indexer_key : bool
+            Whether key is a _LocIndexer key
+        """
+        column_axis = 1
+        if is_indexer_key:
+            if not (
+                isinstance(key, tuple)
+                and len(key) >= 2  # key is at least 2-dimensional
+                and is_list_like_indexer(
+                    key[column_axis]
+                )  # key indexes multiple columns
+                and not com.is_bool_indexer(key[column_axis])
+            ):
+                return
+            key = key[column_axis]
+
+        if not isinstance(self.obj._get_axis(column_axis), ABCMultiIndex) and all(
+            is_hashable(k) for k in key
+        ):
+            for k in key:
+                try:
+                    self.obj[k]
+                except KeyError:
+                    self.obj[k] = np.nan
+
+    def _get_setitem_indexer(self, key):
+        self._ensure_listlike_indexer(key, is_indexer_key=True)
```

Comment:
ideally can you move this to the base class (so don't need to override here)

Reply:
Where should overriding happen then? Define `_ensure_listlike_indexer` in `DataFrame` class too and override `_ensure_listlike_indexer` in `_LocIndexer`?

Reply:
no, I mean move to _NDFrameIndexer (as opposed to _LocIndexer), then you won't have to override, you can just *always* call it in _get_setitem_indexer; that is the goal here.

we want to avoid special casing in the sub-classes of indexers, and allow _ensure_listlike_indxer to be called *always* (it just won't do anything if its conditions are not met)

close on this, thanks for working on it!

Reply:
Sorry, it's a typo. I wanted to say `_NDFrameIndexer` instead of `DataFrame`.

How can we ensure that conditions are only met if indexer is a `_LocIndexer`?